pkg/operators/ebpf/stats.go

```diff
@@ -0,0 +1,505 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpfoperator
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/spf13/viper"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/bpfstats"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	metadatav1 "github.com/inspektor-gadget/inspektor-gadget/pkg/metadata/v1"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	ParamGadgetStatisticsInterval = "statistics-interval"
+	EmitStatsAnn                  = "emitstats"
+	ParamProgramStats             = "programs"
+	StatsDSName                   = "bpfstats"
+	EnableBPFStatsParam           = "enable-bpfstats"
+)
+
+func (o *ebpfOperator) GlobalParams() api.Params {
+	return api.Params{
+		{
+			Key:          EnableBPFStatsParam,
+			Description:  "Enable capturing eBPF stats",
+			DefaultValue: "false",
+			TypeHint:     api.TypeBool,
+			Title:        "Enable eBPF stats",
+		},
+	}
+}
+
+func (o *ebpfOperator) Init(params *params.Params) error {
+	if !params.Get(EnableBPFStatsParam).AsBool() {
+		return nil
+	}
+
+	o.gadgetObjs = make(map[operators.GadgetContext]gadgetObjs)
+
+	// Enable stats collection
+	// TODO: when to disable it?
+	if err := bpfstats.EnableBPFStats(); err != nil {
+		return fmt.Errorf("enabling bpf stats: %w", err)
+	}
+
+	return nil
+}
+
+func (o *ebpfOperator) InstantiateDataOperator(
+	gadgetCtx operators.GadgetContext, paramValues api.ParamValues,
+) (operators.DataOperatorInstance, error) {
+	cfg, ok := gadgetCtx.GetVar("config")
+	if !ok {
+		return nil, fmt.Errorf("missing configuration")
+	}
+	v, ok := cfg.(*viper.Viper)
+	if !ok {
+		return nil, fmt.Errorf("invalid configuration format")
+	}
+
+	emitStatsAnn := v.GetBool(fmt.Sprintf("annotations.operator.%s.%s", o.Name(), EmitStatsAnn))
+	if !emitStatsAnn {
+		return nil, nil
+	}
+
+	if o.gadgetObjs == nil {
+		return nil, fmt.Errorf("bpfstats aren't enabled")
+	}
+
+	var err error
+
+	instance := &ebpfOperatorDataInstance{
+		bpfOperator:  o,
+		done:         make(chan struct{}),
+		programStats: paramValues[ParamProgramStats] == "true",
+	}
+
+	instance.ds, err = gadgetCtx.RegisterDataSource(datasource.TypeArray, StatsDSName)
+	if err != nil {
+		return nil, err
+	}
+
+	intervalAnn := paramValues[ParamGadgetStatisticsInterval]
+	instance.interval, err = time.ParseDuration(intervalAnn)
+	if err != nil {
+		return nil, fmt.Errorf("parsing duration: %w", err)
+	}
+
+	instance.ds.AddAnnotation(api.FetchIntervalAnnotation, intervalAnn)
+	instance.ds.AddAnnotation("cli.clear-screen-before", "true")
+
+	if nodeName, ok := os.LookupEnv("NODE_NAME"); ok {
+		instance.nodeNameField, err = instance.ds.AddField("nodeName", api.Kind_String,
```

Comment:
Or should the description be included here when adding the fields?

Reply:
I think it need to go here where fields are created.